### ISSUE TITLE
"pub global activate <up-to-date-package>" takes 6(?) seconds

### ISSUE THREAD

**Issue by DanTup** [dart-lang/pub#2097]
I'd like to make Dart Code automatically update the Dev Tools package if it's already installed. In IntelliJ this is done by just always running `pub global activate devtools` when launched. However, on my MacBook it takes over 6 seconds to do that *when there are no updates* and that seems like a lot of time to add to every debug session.

Is it expected to take this long? And if so, is there a fast way I can check if it's already up-to-date? I could scrape the pub site to check the latest version, but it feels a bit weird doing it that way, so I'm hoping there's something more "supported".

```
dantup-macbookpro:tmp dantup$ time pub global activate --no-executables devtools
Package devtools is currently active at version 0.0.12.
Resolving dependencies... (4.4s)
+ args 1.5.1
+ async 2.0.8
+ charcode 1.1.2
+ codemirror 0.5.3+5.44.0
+ collection 1.14.11
+ convert 2.1.1
+ devtools 0.0.12 (0.0.13-dev.1 available)
+ http 0.12.0+1
+ http_parser 3.1.3
+ intl 0.15.7
+ js 0.6.1+1
+ meta 1.1.7
+ mime 0.9.6+2
+ octicons_css 0.0.1
+ path 1.6.2
+ pedantic 1.5.0
+ plotly_js 0.0.1
+ polymer_css 0.0.1
+ primer_css 0.0.1
+ rxdart 0.21.0
+ shelf 0.7.4
+ shelf_static 0.2.8
+ source_span 1.5.5
+ split 0.0.1
+ stack_trace 1.9.3
+ stream_channel 1.6.8
+ string_scanner 1.0.4
+ term_glyph 1.1.0
+ typed_data 1.1.6
+ vm_service_lib 3.14.2 (3.14.3-dev.3 available)
Precompiling executables... (1.3s)
Precompiled devtools:devtools.
Activated devtools 0.0.12.

real	0m6.149s
user	0m3.096s
sys	0m0.421s
```

**Comment by jonasfj** [comment]
> Is it expected to take this long?

I don't think it's surprising, but that's not to say that there isn't things we could do to make it faster :)

If no dependencies were updated we might not have to redo: `Precompiling executables... (1.3s)`.

It could also be interesting to look in `Resolving dependencies... (4.4s)` and see how much time is spent fetching a list of available versions for each package, and how much time is spend actually resolving dependencies. My intuition would be that this is dominated by I/O and network round-time, and that it's hopefully already doing most of the requests concurrently. I haven't looked at the code yet, so that's an assumption. A first step would be to play with verbosity, maybe there already is instrumentation for this.

> And if so, is there a fast way I can check if it's already up-to-date?

Depends on your definition of up-to-date :)

If up-to-date means I have the latest version of `devtools`, then yes, you can check that with a quick call to `https://pub.dartlang.org/api/packages/<package>` which will return JSON that includes the list of all known versions of the given `<package>`. This end-point is likely to remain stable for a long time, as it's used by `pub` distributed with Dart/Flutter SDKs (deprecation would be slow).

if up-to-date means, I have the latest version of `devtools` **AND the latest version of all compatible dependencies**, the answer is probably that you won't be able to do it much faster.

If you want to investigate why it's slow, and whether it could be made faster that would be very useful, but I suspect this code-path is pretty tight.

Note. you'll probably want latest compatible version of dependencies in cases where a new version of `pedantic` is available, or a new version of `pedantic` became compatible because dart-sdk on the system was updated.

---
For your use-case I would consider updating in the background, and updating automatically no-more frequently than once every few days (storing the last date updated in a config file somewhere).

**Comment by DanTup** [comment]
> If no dependencies were updated we might not have to redo: `Precompiling executables... (1.3s)`.

As far as I can tell, it's not actually doing anything here - since if I run it again, I get exactly the same time (the one I posted was after running it several times). If it was actually updating anything, then it shouldn't need to do it the next time (in other words, as far as I can tell, there are no actual effects of running this when it takes 6 seconds - I don't think anything was changed).

> For your use-case I would consider updating in the background

That might work actually. I'd prefer to show the user we were updating (so if they see it updating when they try to launch, the might go check the release notes if they want to see it), but that's not the end of the world.

Still, it'd be interesting to know why it's taking 6 seconds. If it doesn't update dependencies (which is doesn't seem to, since I get the same output/timing of subsequent runs) I wonder if it really needs to check them. Here's a verbose log if it's useful (which shows what you might expect - requests for each package, and then a compilation step):

[verbose.txt](https://github.com/dart-lang/pub/files/2952495/verbose.txt)

**Comment by DanTup** [comment]
Slightly related - I noticed that running `pub global activate devtools` will *downgrade me* if I'm on a pre-release version. Is there a way to make it only ever upgrade? If someone is testing a pre-release build, implementing the auto-update by just calling `pub global activate devtools` in the background will downgrade them.

If there's no support for this, then calling the API might be better then, so we can choose only to run it when we know it's an upgrade.

**Comment by jonasfj** [comment]
Hmm, downgrading from a pre-release might be desirable... And an option to not downgrade seems overkill for this particular corner case..

But if we added an command to inspect the installed version, then you could do specify version constraint that it much be greater than what is currently installed..

Adding a `pub global describe <package>` which says if it's installed and what version would probably be uncontroversial..

**Comment by jonasfj** [comment]
But the operation is slow because it fetches a list of all versions available for each dependency. And naturally does so recursively for each dependency that it tries to use..

Like I said, I assume these requests are mostly concurrent, but it's possible they are not.. we should instrument when trying to understand why they are slow..

**Comment by DanTup** [comment]
> downgrading from a pre-release might be desirable

Sure, though to me it seems a bit weird to downgrade in a command that's generally used to upgrade without some sort of flag.

I've pushed a change that uses the API for now for checking updates, so don't worry about doing anything specifically for my case - the things I mentioned seem like they might be generally useful, but that's a decision for you!

> But the operation is slow because it fetches a list of all versions available for each dependency.

Yeah, though I'm not sure why it's doing this if it's not going to do anything with that information. If it's a complete no-op if you're already on the latest version of the specified package, it seems like you could bail out really early?

Again, not a concern for me now, but I don't know if it's doing work it doesn't need to. Feel free to close/re-purpose this as you see fit, and let me know if I can provide any more info (or if, fox ex. you can't repro this performance).

Thanks!

**Comment by jonasfj** [comment]
hmm, maybe you are on to something...
If you grep verbose.txt for `"took"` you see that most of the HTTP requests took `0.13` seconds, but if you add it up, all of those makes it `4.08` seconds used on downloading files... which sums suspiciously close to the `4.48` seconds used for resolving versions.

This leads me to suspect that `pub` fetching lists of versions one-by-one instead of doing it concurrently... This might be hard to fix because the resolver won't load versions until they are needed.
Maybe it's enough to eagerly start loading `_versions` from the constructor in:
https://github.com/dart-lang/pub/blob/master/lib/src/solver/package_lister.dart#L74

I bet this could be made faster, but it's probably not easy...

----
side note, we're building `pub.dartlang.org` to handle serious load and auto-scale, but checking for updates automatically it might be nice to only do it once per day or so. For your use-case it would almost always be bad to update in a blocking path, in case the user has a slow internet connection, etc..

**Comment by jonasfj** [comment]
>> But the operation is slow because it fetches a list of all versions available for each dependency.
>
> Yeah, though I'm not sure why it's doing this if it's not going to do anything with that information.

It needs to upgrade dependencies... if a new version of one of your dependencies have been published it needs to use it.

**Comment by DanTup** [comment]
> This leads me to suspect that pub fetching lists of versions one-by-one instead of doing it concurrently... 

It does look that way. It also makes me wonder if this work could be done on the server... is every client resolving the dependencies itself, with a bunch of HTTP requests? Wouldn't it be nice if the client could ask the server for a package and get the full tree (which could be cached on the server side)? Not only would slashing the number of HTTP requests speed things up, it may also reduce a lot of load on the server if there are a large number of clients.

(I don't actually have any idea how this all works, so this may be nonsense - I'm just thinking aloud!).

> but checking for updates automatically it might be nice to only do it once per day or so

Yeah, that makes sense - I'll make limit it to once per day.

> For your use-case it would almost always be bad to update in a blocking path

Since this is an optional component and VS Code is on a drive to track down performance-sapping extensions, I don't want to be kicking off more stuff in the background during startup (we're already kicking off analysis, flutter daemon, SDK update checks). I think the current flow works well - the update is done in front of the user and when they've choose to do launch the Dev Tools. I suspect in future we'll have a flag to opt-out of auto-updates in which case doing it at launch-time would also feel more natural than another prompt at startup. I also think running `pub global activate` silently in the background immediately at startup may be a bit aggressive.

I'll add the once-per-day change for now, and can always re-visit it based on feedback and/or how it feels to me using it :-)

**Comment by jonasfj** [comment]
> Wouldn't it be nice if the client could ask the server for a package and get the full tree (which could be cached on the server side)?

Yeah, I've been thinking about that too... On the other hand it's nice to have small HTTP requests because that scales better... More importantly this is going to be a problem if you have custom pub servers, local dependencies and private git repositories... So it's probably best to do version resolution locally..

**Comment by DanTup** [comment]
I was thinking if the requests are concurrent, won't the minimum time always be at least number_of_levels * round_trip_time which could be pretty slow for deep trees?

The full resolution info could always be provided optionally, so custom servers don't need to provide it if it's difficult. It seems like a big win though if a package can be downloaded 100,000 times with the same pre-computed resolution info rather than 100,000 machines all doing the work (and waiting on the round trips).

That said, I won't pretend I know a lot about this because I don't :) My original issue is solved now - I'm just using the API and only checking if it's been at least 20 hours since the last check. I do have a lot of CI builds that pull down packages though, so any improvements would be welcomed (I've been improving caching recently, which seems to be helping a lot, especially for the Flutter repo!).

(feel free to close this if you don't think there's anything concrete to do here btw!)

**Comment by jonasfj** [comment]
> I was thinking if the requests are concurrent, won't the minimum time always be at least number_of_levels * round_trip_time which could be pretty slow for deep trees?

Depends on how deep and how slow :)

But I think I outlined (https://github.com/dart-lang/pub/issues/2097#issuecomment-471590003) that requests might not be concurrent. In fact I now strongly suspect they aren't...

If we made requests to list all versions for each dependency of a package we've shown interest during version resolution in concurrent, then I think we would have `number_of_levels * round_trip_time` instead of `number_of_packages * round_trip_time`. Many packages have a lot of dependencies and they often share some of them with their own dependencies, so I doubt it would be very deep.

It's also possible that it could be made even more concurrent than that.. but I think this is very hard to do.
And for the most part resolution is pretty fast.. I think the way to make it faster would be to cache list of package versions locally, and have a way to query if there was updates to any of those packages... No fully thought out plans.. there might be other things to do first..

**Comment by jonasfj** [comment]
@DanTup, if you're checking `pubspec.lock` into your repository and cache `~/.pub-cache` between builds, I think CI will be all local unless there was an update to `pubspec.lock`.

**Comment by DanTup** [comment]
>  if you're checking `pubspec.lock` into your repository

Most of my CI is running other peoples pub packages, or building Flutter. I've got PubCache in the CI builds, and I've also taken to caching the entire Flutter folder (then git fetching over the top) so that its own package lock files are valid when there are no Flutter updates (which skips a ton of work). I've gotten the Flutter restores down from 4-5 minutes to around 50 seconds for Dart Code's integration tests which is a good start :-)
